### DISCUSSION TITLE
Classifying constants in function bodies

### DISCUSSION THREAD

**Discussion by tinygiant98** [tree-sitter/tree-sitter#1553] [Q&A] [answered]
I'm working on a grammar for a custom language that's based on c, so the vast majority of the grammar was c worked out of the gate.  I'm having a little trouble trying to find a method to classify previously identified constants.  For example, if I have the following at the top of a file:
```
const string MY_STRING = "MY_STRING";
```
I can easily classify `const` (modifier), `string` (type), and `MY_STRING` (variable.readonly).  However, later in the file, I'm unable to isolate the use of the defined constants.  Is there a way to to classify `MY_STRING` as variable.readonly in the following code?
```
void myProcedure(int n = 0)
{
    string s = MY_STRING;
}
```
It's easy to classify `string` (type) and `s` (variable.readwrite), but I can't identify `MY_STRING` as variable.readonly.  It gets classified as variable.readwrite.  Thanks!

**Comment by ahelwer** [comment, marked as answer]
Tree-sitter is a parsing library that analyzes code syntax, while this sort of reference resolution/type inference falls pretty squarely over the line into semantic analysis. It's the sort of fact you derive about your code by examining the syntax tree after parsing has completed, rather than determining it at parse time. There's good reason for this; reference resolution can get extremely complicated. Check out [stack graphs](https://github.blog/2021-12-09-introducing-stack-graphs/) if you're interested in pursuing this.

Unless you're talking about highlighting - tree-sitter has built-in capability for reference-based highlighting at the file level. See the documentation [here](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables) and my struggles with it [here](https://github.com/tree-sitter/tree-sitter/discussions/1326). Neovim also doesn't quite yet support reference highlighting with tree-sitter; see [here](https://github.com/nvim-treesitter/nvim-treesitter/discussions/1667).

**Reply by tinygiant98** [reply]
Awesome, thanks for the detailed response. I kind of figured that would be the case. I'm using this in a syntax highlighter, so that's what I'm looking for. I'll follow up on your links. Thanks again!
